### PR TITLE
Package Themis Core with BoringSSL

### PR DESCRIPTION
**Pull request by ilammy**
This PR teaches `make deb`, `make deb_php`, `make rpm` to build packages of Themis Core using embedded BoringSSL instead of system OpenSSL. This is useful in cases like #619, #657.

When `ENGINE=boringssl` is set, produced Themis Core packages will get a suffix:

  - `libthemis-boringssl`
  - `libthemis-boringssl-dev` or `libthemis-boringssl-devel`

This is an alternative flavor of Themis Core which includes embedded BoringSSL instead of depending on system OpenSSL.

The BoringSSL and OpenSSL flavors provide the same ABI and install the same files. As such, they are mutually exclusive and conflict with each other to prevent simultaneous installation.

Other packages (`libthemispp-dev`, `libthemis-jni`, `libphpthemis`) now depend on _either_ OpenSSL or BoringSSL flavor. That is, any flavor can be installed to satisfy the dependency.

### How to build new packages

In short, running

```
make deb ENGINE=boringssl
```

will build and package BoringSSL flavor of Themis. This will also compile and package other packages which are identical to the OpenSSL version.

It's better to build OpenSSL and BoringSSL flavors in separate build directories right now. The build system does not expect incremental rebuilds with different configuration.

### Dependency summary

The dependencies are as follows:

  - `libthemis`
    - Depends: `libsslX.Y` (of any version)
    - Conflicts: `libthemis-boringssl` of any version **(new!)**
  - `libthemis-boringssl` **(new!)**
    - Depends: _none_
    - Conflicts: `libthemis` of any version
  - `libthemis-dev`
    - Depends: `libssl-dev`
    - Depends: `libthemis` of exactly the same version
    - Conflicts:  `libthemis-boringssl-dev` of any version **(new!)**
  - `libthemis-boringssl-dev` **(new!)**
    - Depends: `libthemis-boringssl` of exactly the same version
    - Conflicts:  `libthemis-dev` of any version
  - `libthemispp-dev`
    - Depends: `libthemis-dev` or `libthemis-boringssl-dev` of exactly the same version **(changed)**
  - `libthemis-jni`
    - Depends: `libthemis` or `libthemis-boringssl` of _at least_ the same version **(changed)**
  - `libphpthemis-php*`
    - Depends: `libthemis` or `libthemis-boringssl` of _at least_ the same version **(new!)**

Ditto for RPM, with `s/dev/devel/g`, etc.

## Checklist

- [X] ~~Change is covered by automated tests~~ (not really, only Builtbot)
- [X] The [coding guidelines] are followed
- [X] Changelog is updated

[coding guidelines]: https://github.com/cossacklabs/themis/blob/master/CONTRIBUTING.md

This PR depends on the previous ones:

- [x] #681 Include embedded BoringSSL into libsoter.a (for the package contain usable static libraries)
- [x] #682 More accurate package dependencies (cherry-picked here)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -880,10 +914,11 @@ pkginfo:
 
 PHP_VERSION_FULL:=$(shell php -r "echo PHP_MAJOR_VERSION.'.'.PHP_MINOR_VERSION;" 2>/dev/null)
 ifeq ($(OS_CODENAME),jessie)
-    PHP_DEPENDENCIES:=php5
+    PHP_DEPENDENCIES += --depends php5
 else
-    PHP_DEPENDENCIES:=php$(PHP_VERSION_FULL)
+    PHP_DEPENDENCIES += --depends php$(PHP_VERSION_FULL)
 endif
+PHP_DEPENDENCIES += --depends "$(PACKAGE_NAME) (>= $(VERSION)+$(OS_CODENAME)) | $(OTHER_PACKAGE_NAME) (>= $(VERSION)+$(OS_CODENAME))"
```
Comment (vixentael) [inline Makefile:921]: magic 🧙‍♂️

File: Makefile [hunk 2]
```diff
@@ -604,23 +604,44 @@ endif
 # Packaging Themis Core: Linux distributions
 #
 
+ifeq ($(ENGINE),boringssl)
+ifeq ($(CRYPTO_ENGINE_LIB_PATH),)
+PACKAGE_EMBEDDED_BORINGSSL := yes
```
Comment (vixentael) [inline Makefile:609]: so, if engine is boringssl, we'll pack boringssl as embedded library by default, right?

Comment (ilammy) [reply Makefile:609]: If `ENGINE=boringssl` _**and**_ `ENGINE_LIB_PATH` is not set. That is, if the user has requested a build with BoringSSL from Themis submodule, not with some user-provided BoringSSL build from user-specified location.

File: Makefile [hunk 3]
```diff
@@ -604,23 +604,44 @@ endif
...
 LICENSE_NAME = "Apache License Version 2.0"
 
 DEB_CODENAME := $(shell lsb_release -cs 2> /dev/null)
 DEB_ARCHITECTURE = `dpkg --print-architecture 2>/dev/null`
-DEB_DEPENDENCIES := --depends openssl
+ifneq ($(PACKAGE_EMBEDDED_BORINGSSL),yes)
+# If we were using native Debian packaging, dpkg-shlibdeps could supply us with
+# accurate dependency information. However, we build packages manually, so we
+# use dependencies of "libssl-dev" as a proxy. Typically this is "libssl1.1".
+DEB_DEPENDENCIES += --depends $(shell apt-cache depends libssl-dev | grep 'Depends:' | cut -d: -f 2-)
```
Comment (shadinua) [inline Makefile:623]:
```suggestion
DEB_DEPENDENCIES += --depends $(shell apt-cache depends libssl-dev | grep 'Depends:' | cut -d: -f 2- | tr -d ' ')
```
:)

Comment (ilammy) [reply Makefile:623]: It will be adjusted like this after #682 is merged 👌

**Comment by ilammy** [comment]
Dependencies have been merged, I've synced up this branch and resolved merge conflicts. If you want to make another review pass over the clean diff, this is the time. I'll merge it today if CI passes. (Let's hope GHA will work this time.)

**Comment by ilammy** [comment]
Well, GHA *is* stuck ¯\\\_(ツ)\_/¯

**Pull request merged**